modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java

```diff
@@ -374,21 +374,17 @@ protected NodeChannels connectToChannels(DiscoveryNode node, ConnectionProfile p
         return nodeChannels;
     }
 
-    private class ChannelCloseListener implements ChannelFutureListener {
-
-        private final DiscoveryNode node;
-
-        private ChannelCloseListener(DiscoveryNode node) {
-            this.node = node;
-        }
-
+    private final class ChannelCloseListener implements ChannelFutureListener {
         @Override
         public void operationComplete(final ChannelFuture future) throws Exception {
-            onChannelClosed(future.channel());
-            NodeChannels nodeChannels = connectedNodes.get(node);
-            if (nodeChannels != null && nodeChannels.hasChannel(future.channel())) {
-                threadPool.generic().execute(() -> disconnectFromNode(node, future.channel(), "channel closed event"));
+            try {
+                onChannelClosed(future.channel());
```

Comment:
I know there are are a bunch of checks at various points so that we can call this close methods multiple times safely. But I just want to point out that in this case:

1. We call `onChannelClosed` here
2. In the finally block below we call `disconnectFromNodeChannel` which dispatches a Runnable to a generic thread pool.
3. At the end of `disconnectFromNodeChannel` we call `onChannelClosed` again with the same channel.

That is all probably fine. But this all seems to be pretty complicated to me right now. Maybe we can find some ways to simplify in the future?